src/workerd/io/worker-fs.c++

```diff
@@ -0,0 +1,935 @@
+#include "worker-fs.h"
+
+#include <workerd/io/io-context.h>
+
+#include <algorithm>
+
+namespace workerd {
+
+#define DEFINE_DEFAULTS_FOR_ROOTS(name, path)                                                      \
+  const jsg::Url FsMap::kDefault##name##Path = path##_url;
+KNOWN_VFS_ROOTS(DEFINE_DEFAULTS_FOR_ROOTS)
+#undef DEFINE_DEFAULTS_FOR_ROOTS
+
+namespace {
+// The SymbolicLinkRecursionGuardScope is used on-stack to guard against
+// circular symbolic links. As soon as a cycle is detected, it throws.
+// Since resolution is always synchronous, we can use thread-local to
+// track the current scope, allowing multiple scopes to be in the stack
+// without needed to pass the guard around or do any other bookkeeping.
+class SymbolicLinkRecursionGuardScope;
+thread_local SymbolicLinkRecursionGuardScope* symbolicLinkGuard = nullptr;
+class SymbolicLinkRecursionGuardScope final {
+ public:
+  SymbolicLinkRecursionGuardScope() {
+    if (symbolicLinkGuard == nullptr) {
+      symbolicLinkGuard = this;
+    }
+  }
+  ~SymbolicLinkRecursionGuardScope() noexcept(false) {
+    if (symbolicLinkGuard == this) {
+      symbolicLinkGuard = nullptr;
+    }
+  }
+
+  static void checkSeen(SymbolicLink* link) {
+    if (symbolicLinkGuard == nullptr) {
+      return;
+    }
+    auto& guard = *symbolicLinkGuard;
+    JSG_REQUIRE(guard.linksSeen.find(link) == kj::none, Error, "Recursive symbolic link detected");
+    guard.linksSeen.insert(link);
+  }
+
+ private:
+  kj::HashSet<SymbolicLink*> linksSeen;
+};
+
+// Thread-local storage to track the current temp directory storage scope
+// on the stack.
+static thread_local TmpDirStoreScope* tmpDirStorageScope = nullptr;
```

Comment:
I wonder if this needs to be a `kj::EventLoopLocal`

Reply:
I don't believe so. We don't want it to be tied to a specific event loop. If it's not coming from the IoContext, then it truly is temporary on the current execution scope... for instance, in the top-level bootstrap of worker, or the top-level evaluation of a dynamically imported module, etc. I don't think `kj::EventLoopLocal` achieves anything here. Could be wrong tho.